doc/adr/0008-jetstream-kv.md

```diff
@@ -0,0 +1,319 @@
+# 8. JetStream based Key-Value Stores
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-06-30|
+|Author  |@ripienaar|
+|Status  |Partially Implemented|
+|Tags    |jetstream, client, kv|
+
+## Context
+
+This document describes a design and initial implementation of a JetStream backed key-value store. The initial implementation
+is available in the CLI as `nats kv` with an initial reference client package implementation in `jsm.go` repository to support
+the CLI.
+
+This document aims to guide client developers in implementing this feature in the language clients we maintain.
+
+## Overview
+
+We intend to hit a basic initial feature set as below, with some future facing goals as indicated:
+
+Initial feature list:
+
+ * Multiple named buckets full of keys with n historical values kept per value
+ * Put and Get of `string(k)=[]byte(v)` values
+ * Put only if the sequence of the last value for a key matches an expected sequence
+ * Historic values can be kept per key, setable on creation of the bucket
+ * Key deletes preserves history
+ * Keys can be expired from the bucket based on a TTL
+ * Watching a specific key or the entire bucket for live updates
```

Comment:
Will this allow me to iterate over all present values in the order they where received?

Reply:
The interesting operation that is not currently possible is to get all the last values. - ie, per last message per subject.

Reply:
Since messages are received in order, the last value will be received after it's corresponding historical values. 
I'm looking at it from the perspective of learning about all keys, so for me that would work. 

Only retrieving the last value has some more implications. Say you got Key A through this, but before you are done iterating Key A changes. What happens then?

I wonder, if only getting the last value is important to you, would you even configure them?

Reply:
Of course, because think about how you can use that as a log - for example the historic changes to all JWTs for an account.

Reply:
You got me convinced, skipping old keys would nice.

Reply:
@matthiashanel yeah the cache is built with a bucket watch - it starts at oldest message and gets every message and write/delete/write etc as the entries flow.  Once it reaches pending=0 it knows the entire cache is built and consisten and then it turns from a passthrough cache to a one thats serving results.

So reading entire history - and indeed discovering all keys etc - is an option, but we want to add a API for getting all known subjects to JS since it knows that now and can gie it quickly.